src/v/cloud_topics/l1/metastore.h

```diff
@@ -0,0 +1,104 @@
+ */
+#pragma once
+
+#include "base/seastarx.h"
+#include "cloud_topics/types.h"
+#include "container/fragmented_vector.h"
+#include "model/fundamental.h"
+#include "model/timestamp.h"
+#include "model/topic_revision.h"
+
+#include <seastar/core/future.hh>
+
+#include <expected>
+
+namespace experimental::cloud_topics::l1 {
+
+// Interface to interact with the L1 metastore. Meant to be totally agnostic to
+// the underlying implementation, whether it's an in-memory store, replicated
+// store, or even an external service.
+//
+// The metastore contains metadata about cloud topic partitions and the L1
+// objects in which they store their data. It is designed to be a simple state
+// store, performing atomic updates to state and answering simple queries for
+// callers. The metastore itself isn't meant to perform complex partition
+// storage management tasks (e.g. garbage collection or compaction), but rather
+// it exposes primitives to enable robustly performing such tasks.
+//
+// At a high level, the metastore contains the following metadata:
+// - a fixed set of information about each Kafka partition (e.g. start offset,
+//   next offset),
+// - the set of extents (pointers into L1 objects) associated with each Kafka
+//   partition, with enough metadata to find an L1 object by offset or
+//   timestamp,
+// - the set of L1 objects and some information about them to enable efficient
+//   object downloads (e.g. footer location).
+//
+// While callers should generally avoid sending invalid requests to the
+// metastore (e.g. creating gaps in offset space), it is up to the metastore
+// implementation to ensure such requests are rejected and don't have harmful
+// side effects. As such, callers can think of this interface as thread safe.
+class metastore {
+public:
+    enum class errc {
+        missing_ntp,
+        invalid_request,
+    };
+    struct object_metadata {
+        struct ntp_metadata {
+            model::tpr tpr;
+            kafka::offset base_offset;
+            kafka::offset last_offset;
+            model::timestamp base_timestamp;
+            model::timestamp last_timestamp;
+            first_byte_offset_t pos;
+            byte_range_size_t size;
+        };
+
+        object_id oid;
+        first_byte_offset_t footer_pos;
+        chunked_vector<ntp_metadata> ntp_metas;
+    };
+
+    struct object_response {
+        object_id oid;
+        size_t footer_pos;
+    };
+
+    struct offsets_response {
+        kafka::offset start_offset;
+        kafka::offset next_offset;
+    };
+
+    // Returns offsets (e.g. start, next) for the given partition.
+    virtual ss::future<std::expected<offsets_response, errc>>
+    get_offsets(const model::tpr&) = 0;
+
+    // Adds the given objects to the metastore. If any are invalid, e.g.
+    // because they break an invariant of a partition's offset ranges, all
+    // objects are rejected.
+    virtual ss::future<std::expected<std::monostate, errc>>
```

Comment:
i'm pretty sure with `std::expected` you can use the void specialization `std::expected<void, errc>`.

Reply:
Yeah, but it's actually kinda ugly:

```
co_return std::expected<void, metastore::errc>{};
```

vs just co_return `std::monostate{}`. I can make the change if you want, or maybe I'm missing a cleaner way to co_return...